gen/__init__.py

```diff
@@ -415,18 +423,24 @@ def add_builtin(name, value):
     temporary_str = 'DO NOT USE THIS AS AN ARGUMENT TO OTHER ARGUMENTS. IT IS TEMPORARY'
     add_builtin('expanded_config', temporary_str)
 
-    return [base_source], targets, templates
+    # Note: must come last so the hash of the "base_source" this is beign added to contains all the
+    # variables but this.
+    add_builtin('sources_id', hash_checkout([hash_checkout(source.make_id()) for source in sources]))
```

Comment:
This could be hidden inside a `Source.finalize()` method. That method could set a flag that enforces no added variables after it's been called.